src/gtil/predict.cc

```diff
@@ -19,46 +19,103 @@
 #include "../threading_utils/parallel_for.h"
 #include "./pred_transform.h"
 
+namespace treelite {
+
+class GTILBridge {
+ public:
+  template <typename ThresholdType, typename LeafOutputType>
+  inline static const typename Tree<ThresholdType, LeafOutputType>::Node*
+  GetNode(const Tree<ThresholdType, LeafOutputType>& tree, int nid)  {
+    return &tree.nodes_[nid];
+  }
+};
+
+}  // namespace treelite
+
 namespace {
 
 using treelite::threading_utils::ThreadConfig;
 using PredTransformFuncType = std::size_t (*) (const treelite::Model&, const float*, float*);
 
-template <typename T>
-inline int NextNode(float fvalue, T threshold, treelite::Operator op,
-                    int left_child, int right_child, int default_child) {
-  if (std::isnan(fvalue)) {
-    return default_child;
+/*!
+ * \brief A dense feature vector.
+ */
+class FVec {
+ public:
+  void Init(std::size_t size) {
+    data_.resize(size);
+    missing_.resize(size);
+    std::fill(data_.begin(), data_.end(), std::numeric_limits<float>::quiet_NaN());
+    std::fill(missing_.begin(), missing_.end(), true);
+    has_missing_ = true;
+  }
+  template <typename DMatrixType>
+  void Fill(const DMatrixType* input, std::size_t row_id) {
+    std::size_t feature_count = 0;
+    input->FillRow(row_id, &data_[0]);
+    for (std::size_t i = 0; i < data_.size(); ++i) {
+      if ( !(missing_[i] = std::isnan(data_[i])) ) {
+        ++feature_count;
+      }
+    }
+    has_missing_ = data_.size() != feature_count;
+  }
+  template <typename DMatrixType>
+  void Clear(const DMatrixType* input, std::size_t row_id) {
+    input->ClearRow(row_id, &data_[0]);
+    std::fill(missing_.begin(), missing_.end(), true);
+    has_missing_ = true;
+  }
+  std::size_t Size() const {
+    return data_.size();
+  }
+  float GetFValue(std::size_t i) const {
+    return data_[i];
+  }
+  bool IsMissing(size_t i) const {
+    return missing_[i];
+  }
+  bool HasMissing() const {
+    return has_missing_;
+  }
+
+ private:
+  std::vector<float> data_;
+  std::vector<bool> missing_;
+  bool has_missing_;
+};
+
+template <typename ThresholdType>
+inline int NextNode(float fvalue, ThresholdType threshold, treelite::Operator op, int left_child) {
+  // XGBoost
+  if (op == treelite::Operator::kLT) {
+    return left_child + !(fvalue < threshold);
+  }
+  // LightGBM, sklearn, cuML RF
+  if (op == treelite::Operator::kLE) {
+    return left_child + !(fvalue <= threshold);
   }
   switch (op) {
     case treelite::Operator::kEQ:
-      return (fvalue == threshold) ? left_child : right_child;
-    case treelite::Operator::kLT:
-      return (fvalue < threshold) ? left_child : right_child;
-    case treelite::Operator::kLE:
-      return (fvalue <= threshold) ? left_child : right_child;
+      return left_child + !(fvalue == threshold);
     case treelite::Operator::kGT:
-      return (fvalue > threshold) ? left_child : right_child;
+      return left_child + !(fvalue > threshold);
```

Comment:
You are making the assumption that nodes are contiguous.  I haven't really verified that for treelite nor xgboost ..  Is it possible that one prune some nodes then reuse the pruned node indexes?

Reply:
In Treelite, the frontend reorders nodes so that they are all contiguous.